README
Perfwatcher - Performance Monitoring
======================================
http://perfwatcher.org

About
-----

  Perfwatcher is a frontend for collectd witch show you graph and
  aggregate data over several host.

Licence
-------

  Perfwatcher is a mix of software under several copying policies including MIT and GPLv2 licences).
  Perfwatcher is released under GPLv2 licence.
  Each file can be used with its own licence (check inside the files)

Release notes
-------------

  See the release notes in the doc/ subdirectory


ChangeLog
2015-01-27, Version 2.2.4

- Bugfix again on View creation
- Fixed issue #57 (specify a tab in the URL)
- Add more lines to top process table

2014-11-29, Version 2.2.3

- Bugfix on View Open when there are more than 10 views
- Bugfix on View creation
- Show an error message in Selections when cdsrc does not exist

2014-11-07, Version 2.2.2

- Bugfix on aggregators configuration (again)

2014-10-08, Version 2.2.1

- Bugfix on aggregators configuration
- missing var $debug_enabled

2014-09-11, Version 2.2

- New Import/Export feature
- Collectd sources : new "Auto-detect" option.
- Bugfix in Top ps window on Memory display
- Top Process : keep sort options when clicking on previous/next
- Bugfix in Vu Meter for CPU and MEM
- Developers only : new checks on json_item_datas (less mem / better control)
- Developers only : container option 'sort' now saved in 'container_options'

2014-03-26, Version 2.1

- Bugs fixed when moving/copying nodes in the tree
- New Postgresql support
- Aggregators can be disabled for some Collectd sources
- Developers only : Collectd source moved from datas to new dedicated column 'cdsrc'
- IE users : Copy a graph to the Windows clipboard (See Release Notes for Mozilla and IE)

2014-02-21, Version 2.0

- New configuration table in db
- New tabs system : using markup language
- New node type : selections. Those are empty nodes where you can add tabs (with markup language)
- No more local access to Collectd. Only using remote features (need jsonrpc plugin)
- Possible use of multiple Collectd sources (need to configure 'cdsrc')
- bin/aggregator.conf uses a configurable path to generate aggregator configuration files
- Upgrade to jquery-ui-1.10.3
- Using JQuery-UI portlets
- Added FreeBSD colors for memory graphs (anonymous user request)
- New icon for left tree root
- Removed vu_cpu_meter and load_meter for folders
- New syntax for direct host/tab access : #host_/source/host(/tab)
- Fixed memory definitions for SSC Serv
- 'Processes' tab now has filters
- New menu design
- New clipboard feature to store graph definitions (for copy/paste to selections)
- New zoom on graph. The new one is intuitive.
- Peuplator can now clean the database (remove orphan items)
- Removed "export" broken feature for graphs.
- Disambiguation for licenses in some files
- Removed keyboard navigation in tabs to prevent conflicts with edition in tabs
- Tabs : use jqueryui tabs
- Tabs : new '+' tab to add a tab
- Tabs : custom tabs are sortable (position is saved)
- Grids : use jqgrid plugin
- Layout : use jquery.layout plugin
- Generic plugin view : use jqueryui button()
- Views : use jqgrid plugin to load a view
- Menu : removed useless entries; moved 'views' to the left
- Updated some dialogs (noty now only for notifications)
- Noty updated to version 2.2.2
- Many bugs fixed

2013-08-29, Version 1.3

- Fixed bug #33 (Broken aggregators / an aggregator for "n")
- new URL syntax : #host_<hostname>/<plugin> (replaces undocumented #host_<hostname>)
- New feature : timeline

2013-04-29, Version 1.2

- New feature : in folder tabs, you can now display folder aggregators. Previously only aggregators (for the selected folder) and hosts/plugins could be shown.
- New feature : views (old menu 'views' was renamed as 'display' and there is a new 'view' menu )
- Updated the database creation scripts (no more automatic initialization)
- Apache proxy is no longer needed. Connection to jsonrpc is now done thanks to curl library
- Bugs fixed

2013-03-25, Version 1.1

- Removed obsolete mysql_* function calls (YM - issue #25)
- Reindent php files (YM)
- Use new topps collectd-pw plugin for top process (CF)
- Folder tabs : limited lifetime (CF - issue #21))
- Code and config cleanup (pexio - #22)
- New bin/get_all_children_path executable (YM)

2012-11-06, Version 1.0

Initial version


js/graph.js
/**
 * Common functions
 *
 * PHP version 5
 *
 * LICENSE: This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @category  Monitoring
 * @author    Cyril Feraudet <cyril@feraudet.com>
 * @copyright 2011 Cyril Feraudet
 * @license   http://www.gnu.org/copyleft/lesser.html  LGPL License 2.1
 * @link      http://www.perfwatcher.org/
 **/ 
var click = 0;
time_begin = 0;
time_end = 0;
function nav_init (time_beginf, time_endf) {
    time_begin = time_beginf;
    time_end = time_endf;
    $('.graph_image').each(function () {
        $(this).attr('ssrc', $(this).attr('isrc').replace (/&(begin|end)=[^&]*/g, ''));
        $(this).click(handle_click);
        $(this).attr('navTimeBegin', new Number (time_begin));
        $(this).attr('navTimeEnd', new Number (time_end));
    });
}

function nav_image_repaint (id)
{
    $('#'+id).attr('src', $('#'+id).attr('ssrc') + '&' 
    + 'begin=' + parseInt($('#'+id).attr('navTimeBegin')).toFixed(0)
    + '&'
    + 'end=' + parseInt($('#'+id).attr('navTimeEnd')).toFixed(0)
    );
} 

function nav_time_reset (id ,diff)
{
  $('#'+id).attr('navTimeEnd', new Number ((new Date ()).getTime () / 1000));
  $('#'+id).attr('navTimeBegin', new Number (parseInt($('#'+id).attr('navTimeEnd')) - diff));
  nav_image_repaint (id);
  return (true);
}

function nav_time_change_obj (id, factor_begin, factor_end)
{
  var diff;

    navTimeBegin = parseInt($('#'+id).attr('navTimeBegin'));
    navTimeEnd = parseInt($('#'+id).attr('navTimeEnd'));
    diff = navTimeEnd - navTimeBegin;
    
    /* Prevent zooming in if diff is less than five minutes */
    if ((diff <= 300) && (factor_begin > 0.0) && (factor_end < 0.0))
      return (true);
    
    navTimeBegin += (diff * factor_begin);
    navTimeEnd   += (diff * factor_end);

    $('#'+id).attr('navTimeBegin', navTimeBegin);
    $('#'+id).attr('navTimeEnd', navTimeEnd);

    nav_image_repaint (id);
    
    return (true);
}

function nav_time_change (id, factor_begin, factor_end)
{
  var diff;

  if (id == '*')
  {
    var all_images;
    var i;

    all_images = document.getElementsByTagName ("img");
    for (i = 0; i < all_images.length; i++)
    {
      if (all_images[i].className != "graph_image")
        continue;
    
      nav_time_change_obj (all_images[i].id, factor_begin, factor_end);
    }
  }
  else
  {
    var img;

    nav_time_change_obj (id, factor_begin, factor_end);
  }

  return (true);
} /* nav_time_change */

function nav_move_earlier (img_id)
{
  return (nav_time_change (img_id, -0.2, -0.2));
} /* nav_move_earlier */

function nav_move_later (img_id)
{
  return (nav_time_change (img_id, +0.2, +0.2));
} /* nav_move_later */

function nav_zoom_in (img_id)
{
  return (nav_time_change (img_id, +0.2, -0.2));
} /* nav_zoom_in */

function nav_zoom_out (img_id)
{
  return (nav_time_change (img_id, (-1.0 / 3.0), (1.0 / 3.0)));
} /* nav_zoom_in */

function nav_set_reference (id)
{
    $(".graph_image").each(function() {
        $(this).attr('navTimeBegin', $('#'+id).attr('navTimeBegin'));
        $(this).attr('navTimeEnd', $('#'+id).attr('navTimeEnd'));
        if ($(this).attr('src') != undefined) {
            nav_image_repaint(this.id);
        }
    });
} 

function nav_recenter (e)
{
    var x;
    var y;
    var diff;
    var time_old_center;
    var time_new_center;
    var width;
    var navTimeEnd, navTimeBegin;
    width = parseInt($('#'+e.target.id).css('width')) - 97;
    navTimeBegin = parseInt($('#'+e.target.id).attr('navTimeBegin'));
    navTimeEnd = parseInt($('#'+e.target.id).attr('navTimeEnd'));
    
    x = e.originalEvent.layerX - 66;
    if (!x || (x < 0) || (x > width))
      return;
    
    y = e.originalEvent.layerY;
    if (!y || (y < 30) || (y > 155))
      return;
    
    diff = navTimeEnd - navTimeBegin;
    time_old_center = navTimeBegin + (diff / 2.0);
    time_new_center = navTimeBegin + (x * diff / width);
    
    $('#'+e.target.id).attr('navTimeBegin', navTimeBegin + (time_new_center - time_old_center));
    $('#'+e.target.id).attr('navTimeEnd', navTimeEnd + (time_new_center - time_old_center));
}

function handle_click (e)
{
    click++
    if (click > 1) { click = 0; nav_handle_dblclick (e); }
    setTimeout(function () {
        if (click == 1) {
            click = 0;
            nav_handle_click (e);
        }
        click = 0;
    }, 300);
}

function nav_handle_dblclick (e)
{
    nav_recenter (e);
    nav_image_repaint (e.target.id)
}

var selecttimespan = 0;
var selecttimespanfunc = false;
var selecttimespangraph = "";

function nav_handle_click (e)
{
    if (selecttimespan == 0) {
        selecttimespan = 1;
        $('#selecttimespan').css('position', 'absolute');
        $('#selecttimespan').css('top', $(e.currentTarget).offset().top + 31);
        if (e.pageX - $(e.currentTarget).offset().left - 67 < 0) {
             $('#selecttimespan').css('left', $(e.currentTarget).offset().left + 67);
        } else if (e.pageX - $(e.currentTarget).offset().left - 67 + 80 > 602) {
             $('#selecttimespan').css('left', $(e.currentTarget).offset().left + 67 + 602 - 80);
        } else {
            $('#selecttimespan').css('left', e.clientX);
        }
        $('#selecttimespan').css('width', 80);
        $('#selecttimespan').resizable({
            autoHide:   true,
            minHeight:  122,
            maxHeight:  122,
            maxWidth:   (602 - (e.pageX - $(e.currentTarget).offset().left - 67)),
            minWidth:   20,
            stop: function(e2, ui) {
                $('#selecttimespan').draggable('option', 'containment', [
                    $(e.currentTarget).offset().left + 66,
                    $(e.currentTarget).offset().top + 30,
                    $(e.currentTarget).offset().left + 669 - parseInt($('#selecttimespan').css('width')),
                    $(e.currentTarget).offset().top + 30
                ]
                );
            }
        });
        $('#selecttimespan').draggable({
            axis: "x",
            appendTo: e.currentTarget,
            containment: [
                $(e.currentTarget).offset().left + 66,
                $(e.currentTarget).offset().top + 30,
                $(e.currentTarget).offset().left + 669 - parseInt($('#selecttimespan').css('width')),
                $(e.currentTarget).offset().top + 30
            ],
            stop: function() {
                $('#selecttimespan').resizable('option', 'maxWidth', 
                $(e.currentTarget).offset().left 
                + e.currentTarget.clientWidth 
                - parseInt($('#selecttimespan').css('left')) 
                -29
                );
            }
        });
        if (selecttimespangraph != '' && selecttimespangraph != e.currentTarget.id) {
            $('#selecttimespan').off('dblclick');
            selecttimespanfunc = false;
        }
        if (!selecttimespanfunc) {
            selecttimespanfunc = true;
            selecttimespangraph = e.currentTarget.id;
            $('#selecttimespan').dblclick(function(){
                var navtimebegin = parseInt($('#'+e.currentTarget.id).attr('navtimebegin'));
                var navtimeend = parseInt($('#'+e.currentTarget.id).attr('navtimeend'));
                var timespan = navtimeend - navtimebegin;
                var width = parseInt($(this).css('width'));
                var start = Math.ceil(parseInt($('#selecttimespan').css('left')) - $(e.currentTarget).offset().left - 66) ;
                var end = Math.ceil(parseInt($('#selecttimespan').css('left')) - $(e.currentTarget).offset().left - 66 + width);
                var offsetbegin = timespan / 602 * start;
                var offsetend = timespan / 602 * end;
                if (new Number(Math.ceil(navtimebegin + offsetbegin + 60)) < new Number(Math.ceil(navtimebegin + offsetend))) {
                    $('#'+e.currentTarget.id).attr('navtimebegin', new Number(Math.ceil(navtimebegin + offsetbegin)));
                    $('#'+e.currentTarget.id).attr('navtimeend', new Number(Math.ceil(navtimebegin + offsetend)));
                    nav_image_repaint (e.currentTarget.id);
                }
                selecttimespan = 0;
                $('#selecttimespan').hide();
            });
        }
        $('#selecttimespan').show();
    } else { 
        selecttimespan = 0;
        $('#selecttimespan').hide();

    }
}

function nav_custom_date(id) {
  $("#date"+id).datepicker({
        onSelect: function(dateText, inst) {
            $('#'+id).attr('navTimeBegin', new Number (dateText / 1000));
            $('#'+id).attr('navTimeEnd', new Number (parseInt($('#'+id).attr('navTimeBegin')) + 86400));
            nav_image_repaint (id);
            $("#date"+id).datepicker("destroy");
        },
        dateFormat: '@',
        showOtherMonths: true,
        selectOtherMonths: true,
        numberOfMonths: 4,
        showButtonPanel: true,
        maxDate: '+0d',
        gotoCurrent: true
     }); 
  return (true);
}

function nav_toggle_legend (img_id) {
    // &tinylegend=1
    img = document.getElementById (img_id);
    var pattern = /tinylegend/;
    if(pattern.test(img.src)) {
        img.src = img.navBaseURL;
    } else {
        img.src = img.navBaseURL + '&tinylegend=1';
    }
}

function add_graph(parenttab, host, plugin_name, plugin_instance_name, type_name, type_instance_name, graph_id) {
    var graph_float = "<div class=\"graph_container\"><div class=\"graph_float\" \
                  parenttab=\""+parenttab+"\" \
                  plugin_name=\""+plugin_name+"\" \
                  plugin_instance_name=\""+plugin_instance_name+"\" \
                  type_name=\""+type_name+"\" \
                  type_instance_name=\""+type_instance_name+"\" \
                > \
                  <img id=\"graph"+graph_id+"\" class=\"graph_image\" \
                    isrc=\"graph.php?host="+host+"&amp;plugin="+plugin_name+"&amp;plugin_instance="+plugin_instance_name+"&amp;type="+type_name+"&amp;type_instance="+type_instance_name+"&amp;begin=-86400&amp;width=500&amp;"+Math.round((new Date()).getTime() / 1000)+"\" \
                    host=\""+host+"\" \
                    parenttab=\""+parenttab+"\" \
                    plugin_name=\""+plugin_name+"\" \
                    plugin_instance_name=\""+plugin_instance_name+"\" \
                    type_name=\""+type_name+"\" \
                    type_instance_name=\""+type_instance_name+"\" \
                  /> \
                  <div class=\"controls zoom\"> \
                    <div title=\"Earlier\" onclick=\"nav_move_earlier ('graph"+graph_id+"');\">&#x2190;</div> \
                    <div title=\"Zoom out\" onclick=\"nav_zoom_out ('graph"+graph_id+"');\">-</div> \
                    <div title=\"Zoom in\" onclick=\"nav_zoom_in ('graph"+graph_id+"');\">+</div> \
                    <div title=\"Later\" onclick=\"nav_move_later ('graph"+graph_id+"');\">&#x2192;</div> \
                  </div> \
                  <div class=\"controls preset\"> \
                    <div title=\"Show current hour\" onclick=\"nav_time_reset ('graph"+graph_id+"', 3600);\">H</div> \
                    <div title=\"Show current day\" onclick=\"nav_time_reset ('graph"+graph_id+"', 86400);\">D</div> \
                    <div title=\"Show current week\" onclick=\"nav_time_reset ('graph"+graph_id+"', 7 * 86400);\">W</div> \
                    <div title=\"Show current month\" onclick=\"nav_time_reset ('graph"+graph_id+"', 31 * 86400);\">M</div> \
                    <div title=\"Show current year\" onclick=\"nav_time_reset ('graph"+graph_id+"', 366 * 86400);\">Y</div> \
                    <div title=\"Custom date\" onclick=\"nav_custom_date ('graph"+graph_id+"');\">C</div> \
                    <div title=\"Set all graph to this timespan\" onclick=\"nav_set_reference ('graph"+graph_id+"');\">!</div> \
                  </div> \
        </div> \
        <div id=\"dategraph"+graph_id+"\"></div> </div>";
        $(graph_float).appendTo('#'+parenttab);
        $('#graph' + graph_id).attr('src', $('#graph' + graph_id).attr('isrc'));
        $('#graph' + graph_id).attr('ssrc', $('#graph' + graph_id).attr('isrc'));
        $('#graph' + graph_id).attr('navTimeBegin', time_begin);
        $('#graph' + graph_id).attr('navTimeEnd', time_end);
        $('#graph' + graph_id).click(handle_click);
        $('#graph' + graph_id).contextMenu({
            menu: 'graphmenu'
        },
            function(action, el, pos) {
            switch(action) {
                case 'top':
                    var timestamp = 0;
                    var navtimebegin = new Number($(el).attr('navtimebegin'));
                    var navtimeend = new Number($(el).attr('navtimeend'));
                    var host = $(el).attr('host');
                    if (pos.x < 67) {
                        timestamp = navtimebegin;
                    } else if (pos.x > 677) {
                        timestamp = navtimeend;
                    } else {
                        timestamp = Math.ceil(navtimebegin + ((navtimeend - navtimebegin) / 600 * (pos.x - 67)));
                    }
                    $.get('index.php?tpl=top&host='+host+'&timestamp='+timestamp, function(data) {
                        $('#topdiv').html(data);
                        $('#topdiv').dialog({ height: 530, width: 530 });
                    });
                break;
                default:
                    alert('Available soon ...');
                break;
            }
        });

}

function updatetop(host, timestamp, sort) {
    $.get('index.php?tpl=top&host='+host+'&timestamp='+timestamp+'&sort='+sort, function(data) {
        $('#topdiv').html(data);
    });
}

/* vim: set sw=4 sts=4 et : */
